lib/aur-fetch

```diff
@@ -8,88 +8,77 @@ readonly startdir=$PWD
 
 # default options
-mode=git recurse=0
+log=directory mode=git recurse=0
 
 git_stat_patch() {
     git --no-pager log -O "$1" --patch --stat '..@{upstream}'
 }
 
-merge_upstream() {
-    # set destination of patch file
-    local pkg=$1 log_dir=$2 output
-
-    if [[ -d $log_dir ]]; then
-        output=$log_dir/$pkg.patch
-    else
-        output=/dev/stdout
-    fi
-
-    # default to showing PKGBUILD diffs first (#399)
-    local orderfile=${GIT_DIR:-$PWD/.git}/orderfile
+upstream_sync() {
+    local pkg=$1 patch=${2:-/dev/stdout}
+    local orderfile=${GIT_DIR:-$PWD/$pkg/.git}/orderfile
 
+    # default to showing PKGBUILD in diffs first (#399)
     if [[ ! -s $orderfile ]]; then
         echo PKGBUILD >"$orderfile"
     fi
 
+    # get last upstream revision
     git fetch -v >&2
 
-    # only print log on upstream changes
     if [[ $(git rev-parse HEAD) != $(git rev-parse '@{upstream}') ]]; then
-        git_stat_patch "$orderfile" >> "$output"
+        # only print log on upstream changes
+        git_stat_patch "$orderfile" >> "$patch"
 
         # discard any local changes (#349)
         git reset --hard "HEAD@{upstream}" >&2
     fi
 }
 
 fetch_git() {
-    local log_dir=$1 pkg uri
+    local uri pkg output_dir=$1
 
     while IFS= read -r uri; do
         pkg=${uri##*/}   # strip path
         pkg=${pkg%%.git} # strip .git suffix
 
         if [[ -d $pkg/.git ]]; then
             # avoid issues with filesystem boundaries (#274)
-            GIT_DIR="$PWD/$pkg"/.git GIT_WORK_TREE="$PWD/$pkg" merge_upstream "$pkg" "$log_dir"
+            GIT_DIR="$PWD/$pkg"/.git GIT_WORK_TREE="$PWD/$pkg" upstream_sync "$pkg" "${output_dir:+$output_dir/$pkg.patch}"
         else
             git clone "$uri"
         fi
     done
 }
 
-diff_log() {
-    local base=$1 log_dir=$2
-
-    if [[ -d $log_dir ]]; then
-        tee -a "$log_dir/$base".diff | diffstat -Ckq -f3
-    else
-        tee
-    fi
-}
-
 tar_no_mode_diff() {
-    local archive=$1 base=$2 dir=$3
-
-    if [[ -d $dir/$base ]]; then
-        LC_MESSAGES=C tar -df "$archive" -C "$dir" | grep -Evq '(Mode|Uid|Gid)'
+    if [[ -d $1/$2 ]]; then
+        LC_MESSAGES=C tar -df "$3" -C "$1" | grep -Evq '(Mode|Uid|Gid)'
     else
         return 1
     fi
 }
 
+diff_log() {
+    tee -a "$1/$2".diff | diffstat -Ckq -f3
+}
+
 fetch_tar() {
     #global tmp
-    local log_dir=$1 archive base
+    local archive base
 
     wget -i /dev/stdin --directory-prefix "$tmp" || return 1
 
     for archive in "$tmp"/*.tar.gz; do
         base=$(basename "$archive" .tar.gz)
         tar -kxf "$archive" --directory "$tmp"
 
-        if tar_no_mode_diff "$archive" "$base" "$startdir"; then
-            diff -ur "$startdir/$base" "$tmp/$base" | diff_log "$base" "$log_dir"
+        if tar_no_mode_diff "$startdir" "$base" "$archive"; then
+            if [[ -d $1 ]]; then
+                diff -ur "$startdir/$base" "$tmp/$base" | diff_log "$1" "$base"
```

Comment:
diffstat is going to stdout. Is this intended?

Reply:
In this case yes, since `diffstat` will only print a summary to `stdout`.

Reply:
Does https://github.com/AladW/aurutils/pull/525/commits/ebbdc5083f635cc3beea9f9aee1f1944c3d75cd9 make it more clear?

Reply:
IMHO, I think it does.

Reply:
:+1: